### PR TITLE
luminous: mgr/balancer: restrict automatic balancing to specific weekdays

### PR DESCRIPTION
**Pull request by k0ste**
We have implemented similar semantics for the scrub behaviour.
Let's try to keep pace with that.

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>
(cherry picked from commit 0412a51ac6dfecc1a72b654196ecb1a599dcc94d)

Conflicts:
- path: src/pybind/mgr/balancer/module.py
  comment: 'get_module_option()' -> 'get_config()' for mgr config-key

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
@k0ste Thanks. Can you include https://github.com/ceph/ceph/pull/26505/commits/b9da0925490d4e9e964249c54e7775d9c76b1719 too?

**Comment by k0ste** [comment]
@xiexingguo in this PR or in another?

**Comment by xiexingguo** [comment]
> @xiexingguo in this PR or in another?

I'd love to do it here.
Thanks!

**Comment by k0ste** [comment]
> I'd love to do it here.
> Thanks!

Cherry-picked!

File: src/pybind/mgr/balancer/module.py [hunk 1]
```diff
@@ -377,25 +377,46 @@ def shutdown(self):
...
+            permit = begin_time <= time_of_day < end_time
         else:
-            return tod >= begin or tod < end
+            permit = time_of_day >= begin_time or time_of_day < end_time
+        if not permit:
+            self.log.debug("should run between %s - %s, now %s, skipping",
+                           begin_time, end_time, time_of_day)
+            return False
+
+        begin_weekday = self.get_config('begin_weekday') or 0
```
Comment (xiexingguo) [inline src/pybind/mgr/balancer/module.py]: @s0nea Seems the old `self.get_config('begin_weekday')` will simply return a string, you need to int(self.get_config('begin_weekday')) first

File: src/pybind/mgr/balancer/module.py [hunk 2]
```diff
@@ -377,25 +377,46 @@ def shutdown(self):
...
+        end_weekday = self.get_config('end_weekday') or 7
```
Comment (xiexingguo) [inline src/pybind/mgr/balancer/module.py]: here too

**Comment by yuriw** [comment]
[wip-yuri3-testing-2019-02-21-2350-luminous](https://trello.com/c/rVcrtd6V/545-wip-yuri3-testing-2019-02-21-2350-luminous)

File: src/pybind/mgr/balancer/module.py [hunk 3]
```diff
@@ -377,25 +377,46 @@ def shutdown(self):
...
+            permit = begin_time <= time_of_day < end_time
         else:
-            return tod >= begin or tod < end
+            permit = time_of_day >= begin_time or time_of_day < end_time
+        if not permit:
+            self.log.debug("should run between %s - %s, now %s, skipping",
+                           begin_time, end_time, time_of_day)
+            return False
+
+        begin_weekday = int(self.get_config('begin_weekday')) or 0
```
Comment (xiexingguo) [inline src/pybind/mgr/balancer/module.py]: @k0ste 
See a lot of errors:
```
 7fcfc96a9700 -1 balancer.serve:
2019-02-22T13:33:35.456 INFO:tasks.ceph.mgr.x.smithi098.stderr:2019-02-22 13:33:35.453750 7fcfc96a9700 -1 Traceback (most recent call last):
2019-02-22T13:33:35.456 INFO:tasks.ceph.mgr.x.smithi098.stderr:  File "/usr/lib/ceph/mgr/balancer/module.py", line 419, in serve
2019-02-22T13:33:35.456 INFO:tasks.ceph.mgr.x.smithi098.stderr:    if self.active and self.time_permit():
2019-02-22T13:33:35.456 INFO:tasks.ceph.mgr.x.smithi098.stderr:  File "/usr/lib/ceph/mgr/balancer/module.py", line 397, in time_permit
2019-02-22T13:33:35.456 INFO:tasks.ceph.mgr.x.smithi098.stderr:    begin_weekday = int(self.get_config('begin_weekday')) or 0
2019-02-22T13:33:35.457 INFO:tasks.ceph.mgr.x.smithi098.stderr:TypeError: int() argument must be a string or a number, not 'NoneType'

```

Should be:
```python
begin_weekday = int(self.get_config('begin_weekday', 0))
end_weekday = int(self.get_config('end_weekday', 7))

Comment (k0ste) [reply src/pybind/mgr/balancer/module.py]: Updated

**Pull request merged**